cmd/oras/internal/option/remote.go

```diff
@@ -0,0 +1,139 @@
+	"os"
+	"strings"
+
+	"github.com/spf13/pflag"
+	oremote "oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"oras.land/oras/internal/credential"
+	"oras.land/oras/internal/http"
+	"oras.land/oras/internal/trace"
+	"oras.land/oras/internal/version"
+)
+
+// Remote options struct.
+type Remote struct {
+	CACertFilePath    string
+	PlainHTTP         bool
+	Insecure          bool
+	Configs           []string
+	Username          string
+	PasswordFromStdin bool
+	Password          string
+}
+
+// ApplyFlags applies flags to a command flag set.
+func (remote *Remote) ApplyFlags(fs *pflag.FlagSet) {
+	fs.StringArrayVarP(&remote.Configs, "config", "c", nil, "auth config path")
+	fs.StringVarP(&remote.Username, "username", "u", "", "registry username")
+	fs.StringVarP(&remote.Password, "password", "p", "", "registry password or identity token")
+	fs.BoolVarP(&remote.PasswordFromStdin, "password-stdin", "", false, "read password or identity token from stdin")
+	fs.BoolVarP(&remote.Insecure, "insecure", "k", false, "allow connections to SSL registry without certs")
+	fs.StringVarP(&remote.CACertFilePath, "ca-file", "", "", "server certificate authority file for the remote registry")
+	fs.BoolVarP(&remote.PlainHTTP, "plain-http", "", false, "allow insecure connections to registry without SSL")
+}
+
+// ReadPassword tries to read password with optional cmd prompt.
+func (remote *Remote) ReadPassword() (err error) {
+	if remote.Password != "" {
+		fmt.Fprintln(os.Stderr, "WARNING! Using --password via the CLI is insecure. Use --password-stdin.")
+	} else if remote.PasswordFromStdin {
+		// Prompt for credential
+		password, err := io.ReadAll(os.Stdin)
+		if err != nil {
+			return err
+		}
+		remote.Password = strings.TrimSuffix(string(password), "\n")
+		remote.Password = strings.TrimSuffix(remote.Password, "\r")
+	}
+	return nil
+}
+
+// tlsConfig assembles the tls config.
+func (remote *Remote) tlsConfig() (config *ctls.Config, err error) {
+	config = &ctls.Config{}
+	var caPool *x509.CertPool
+	if remote.CACertFilePath == "" {
+		caPool = nil
+	} else if caPool, err = http.LoadCertPool(remote.CACertFilePath); err != nil {
+		return nil, err
+	}
+
+	config.RootCAs = caPool
+	config.InsecureSkipVerify = remote.Insecure
+	return
+}
+
+// AuthClient assembles a oras auth client
+func (remote *Remote) AuthClient(debug bool) (client *auth.Client, err error) {
+	config, err := remote.tlsConfig()
+	if err != nil {
+		return nil, err
+	}
+	client = &auth.Client{
+		Client: &nhttp.Client{
+			Transport: &nhttp.Transport{
+				TLSClientConfig: config,
+			},
+		},
+	}
+	client.SetUserAgent("oras/" + version.GetVersion())
+	if debug {
+		client.Client.Transport = trace.NewTransport(client.Client.Transport)
+	}
+
+	cred := credential.Credential(remote.Username, remote.Password)
+	if cred != auth.EmptyCredential {
+		client.Credential = func(ctx context.Context, s string) (auth.Credential, error) {
+			return remote.Credential(), nil
```

Comment:
It is unnecessary to re-compute the `cred`.
```suggestion
			return cred, nil
```